tests/pytorch/test_optim.py

```diff
@@ -42,5 +45,173 @@ def test_sparse_adam():
+    torch_emb_zero = th.nn.Embedding(num_embs, emb_dim, sparse=True)
+    th.manual_seed(0)
+    th.nn.init.uniform_(torch_emb.weight, 0, 1.0)
+    th.nn.init.uniform_(torch_emb_zero.weight, 0, 1.0)
+    th.manual_seed(0)
+    th.nn.init.uniform_(dgl_emb.emb_tensor, 0, 1.0)
+    th.nn.init.uniform_(dgl_emb_zero.emb_tensor, 0, 1.0)
+
+    dgl_adam = SparseAdam(params=[dgl_emb, dgl_emb_zero], lr=0.01)
+    torch_adam = th.optim.SparseAdam(
+        list(torch_emb.parameters()) + list(torch_emb_zero.parameters()), lr=0.01)
+
+    # first step
+    idx = th.randint(0, num_embs, size=(4,))
+    dgl_value = dgl_emb(idx, device).to(th.device('cpu'))
+    torch_value = torch_emb(idx)
+    labels = th.ones((4,)).long()
+
+    dgl_adam.zero_grad()
+    torch_adam.zero_grad()
+    dgl_loss = th.nn.functional.cross_entropy(dgl_value, labels)
+    torch_loss = th.nn.functional.cross_entropy(torch_value, labels)
+    dgl_loss.backward()
+    torch_loss.backward()
+
+    dgl_adam.step()
+    torch_adam.step()
+    assert F.allclose(dgl_emb.emb_tensor, torch_emb.weight)
+
+def initializer(emb):
+    th.manual_seed(0)
+    emb.uniform_(-1.0, 1.0)
+    return emb
+
+def start_sparse_adam_worker(rank, world_size, num_embs=128, emb_dim=10):
+    print('start sparse worker for adam {}'.format(rank))
+    dist_init_method = 'tcp://{master_ip}:{master_port}'.format(
+        master_ip='127.0.0.1', master_port='12345')
+    backend = 'gloo'
+    device=F.ctx()
+
+    th.distributed.init_process_group(backend=backend,
+                                      init_method=dist_init_method,
+                                      world_size=world_size,
+                                      rank=rank)
+
+    dgl_emb = NodeEmbedding(num_embs, emb_dim, 'test', init_func=initializer)
+    torch_emb = th.nn.Embedding(num_embs, emb_dim, sparse=True)
+    th.manual_seed(0)
+    th.nn.init.uniform_(torch_emb.weight, -1.0, 1.0)
+    torch_emb = th.nn.parallel.DistributedDataParallel(torch_emb)
+    dgl_adam = SparseAdam(params=[dgl_emb], lr=0.01)
+    torch_adam = th.optim.SparseAdam(list(torch_emb.module.parameters()), lr=0.01)
+
+    start = (num_embs // world_size) * rank
+    end = (num_embs // world_size) * (rank + 1)
+    idx = th.randint(start, end, size=(4,))
+    dgl_value = dgl_emb(idx, device).to(th.device('cpu'))
+    torch_value = torch_emb(idx)
+    labels = th.ones((4,)).long()
+
+
+    dgl_adam.zero_grad()
+    torch_adam.zero_grad()
+    dgl_loss = th.nn.functional.cross_entropy(dgl_value, labels)
+    torch_loss = th.nn.functional.cross_entropy(torch_value, labels)
+    dgl_loss.backward()
+    torch_loss.backward()
+
+    dgl_adam.step()
+    torch_adam.step()
+    if rank == 0:
+        after_step = dgl_emb(idx, device)
+        assert F.allclose(dgl_emb.emb_tensor, torch_emb.module.weight)
+        assert F.allclose(dgl_value, after_step) is False
+    th.distributed.barrier()
+
+def start_sparse_adam_worker_zero_step(rank, world_size, num_embs=128, emb_dim=10):
```

Comment:
what's the difference between this one and the one above?
also can we merge it with the one above. it seems most of the code is the same.